examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+              << " (" << n_u << '+' << n_p << ')'
+              << std::endl;
+
+    {
+      BlockDynamicSparsityPattern csp (2,2);
+
+      csp.block(0,0).reinit (n_u, n_u);
+      csp.block(1,0).reinit (n_p, n_u);
+      csp.block(0,1).reinit (n_u, n_p);
+      csp.block(1,1).reinit (n_p, n_p);
+
+      csp.collect_sizes();
+
+      DoFTools::make_sparsity_pattern (dof_handler, csp, constraints, false);
+      sparsity_pattern.copy_from (csp);
+
+    }
+    system_matrix.reinit (sparsity_pattern);
+
+    solution.reinit (2);
+    solution.block(0).reinit (n_u);
+    solution.block(1).reinit (n_p);
+    solution.collect_sizes ();
+
+    system_rhs.reinit (2);
+    system_rhs.block(0).reinit (n_u);
+    system_rhs.block(1).reinit (n_p);
+    system_rhs.collect_sizes ();
+  }
+
+
+// @sect4{StokesProblem::assemble_system}
+
+// In this function, the system matrix is assembled the same regardless of using ILU and GMG
+  template <int dim>
+  void StokesProblem<dim>::assemble_system ()
+  {
+    TimerOutput::Scope assemble(computing_timer, "Assemble");
+    system_matrix=0;
+    system_rhs=0;
+
+    double mass_factor = (solver_type == SolverType::UMFPACK) ? 0.0 : 1.0;
+
+    QGauss<dim>   quadrature_formula(degree+2);
+
+    FEValues<dim> fe_values (fe, quadrature_formula,
+                             update_values    |
+                             update_quadrature_points  |
+                             update_JxW_values |
+                             update_gradients);
+
+    const unsigned int   dofs_per_cell   = fe.dofs_per_cell;
+
+    const unsigned int   n_q_points      = quadrature_formula.size();
+
+    FullMatrix<double>   local_matrix (dofs_per_cell, dofs_per_cell);
+    Vector<double>       local_rhs (dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+    const RightHandSide<dim>          right_hand_side;
+    std::vector<Vector<double> >      rhs_values (n_q_points,
+                                                  Vector<double>(dim+1));
+
+    const FEValuesExtractors::Vector velocities (0);
+    const FEValuesExtractors::Scalar pressure (dim);
+
+    std::vector<SymmetricTensor<2,dim> > symgrad_phi_u (dofs_per_cell);
+    std::vector<double>                  div_phi_u   (dofs_per_cell);
+    std::vector<double>                  phi_p       (dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      {
+        fe_values.reinit (cell);
+        local_matrix = 0;
+        local_rhs = 0;
+
+        right_hand_side.vector_value_list(fe_values.get_quadrature_points(),
+                                          rhs_values);
+
+        for (unsigned int q=0; q<n_q_points; ++q)
+          {
+            for (unsigned int k=0; k<dofs_per_cell; ++k)
+              {
+                symgrad_phi_u[k] = fe_values[velocities].symmetric_gradient (k, q);
```

Comment:
of course you are right. I missed the fact that it's a `symmetric_gradient` of a vector field. Nonetheless, i still think it would not make any difference for a single-block problems and values/gradients of scalar field as those things are already stored inside `FEValues` object in the ready-to-use manner. In this case there will be, supposedly, an overhead for getting `fe_values[pressure].value (k, q);` due to extractors.